server/src/test/java/org/elasticsearch/cluster/routing/allocation/ShardsHealthIndicatorServiceTests.java

```diff
@@ -0,0 +1,310 @@
+    ) {
+        return new SimpleHealthIndicatorDetails(
+            Map.of(
+                "allocated_primaries_count",
+                allocatedPrimaries.size(),
+                "allocated_replicas_count",
+                allocatedReplicas.size(),
+                "unreplicated_primaries_count",
+                unreplicatedPrimaries.size(),
+                "unreplicated_primaries",
+                unreplicatedPrimaries.stream().map(it -> it.shards().get(1).shardId()).toList(),
+                "unallocated_replicas_count",
+                unallocatedReplicas.size(),
+                "unallocated_replicas",
+                unallocatedReplicas.stream().map(it -> it.shards().get(1).shardId()).toList(),
+                "restarting_replicas_count",
+                restartingReplicas.size(),
+                "restarting_replicas",
+                restartingReplicas.stream().map(it -> it.shards().get(1).shardId()).toList(),
+                "unallocated_primaries_count",
+                unallocatedPrimaries.size(),
+                "unallocated_primaries",
+                unallocatedPrimaries.stream().map(it -> it.shards().get(1).shardId()).toList()
+            )
+        );
+    }
+
+    private static ClusterState createClusterStateWith(List<IndexRoutingTable> indexes) {
+        var builder = RoutingTable.builder();
+        for (IndexRoutingTable index : indexes) {
+            builder.add(index);
+        }
+        return ClusterState.builder(new ClusterName("test-cluster"))
+            .routingTable(builder.build())
+            .metadata(Metadata.builder().putCustom(NodesShutdownMetadata.TYPE, new NodesShutdownMetadata(Map.of())).build())
+            .build();
+    }
+
+    private static Supplier<IndexRoutingTable> indexGenerator(String prefix, ShardState primaryState, ShardState... replicaStates) {
+        var index = new AtomicInteger(0);
+        return () -> index(prefix + index.incrementAndGet(), primaryState, replicaStates);
+    }
+
+    private static IndexRoutingTable index(String name, ShardState primaryState, ShardState... replicaStates) {
+        var index = new Index(name, UUID.randomUUID().toString());
+        var shardId = new ShardId(index, 1);
+
+        var builder = IndexRoutingTable.builder(index);
+        builder.addShard(createShardRouting(shardId, true, primaryState));
+        for (ShardState replicaState : replicaStates) {
+            builder.addShard(createShardRouting(shardId, false, replicaState));
+        }
+        return builder.build();
+    }
+
+    private static ShardRouting createShardRouting(ShardId shardId, boolean primary, ShardState state) {
+        var routing = newUnassigned(
+            shardId,
+            primary,
+            primary ? RecoverySource.EmptyStoreRecoverySource.INSTANCE : RecoverySource.PeerRecoverySource.INSTANCE,
+            new UnassignedInfo(UnassignedInfo.Reason.INDEX_CREATED, null)
+        );
+        if (state == UNASSIGNED) {
```

Comment:
WDYT about leveraging switch expressions here?

```java
return switch (state) {
    case UNASSIGNED -> routing;
    case STARTED -> routing.initialize(UUID.randomUUID().toString(), null, 0).moveToStarted();
    case UNASSIGNED_RESTARTING -> routing.initialize(UUID.randomUUID().toString(), null, 0)
        .moveToStarted()
        .moveToUnassigned(
            new UnassignedInfo(
                UnassignedInfo.Reason.NODE_RESTARTING,
                null,
                null,
                -1,
                0,
                0,
                false,
                UnassignedInfo.AllocationStatus.DELAYED_ALLOCATION,
                Set.of(),
                UUID.randomUUID().toString()
            )
        );
};
```

Reply:
I am not sure if that is easily achievable. I would say this is not having exclusive branches but rather executing subset of branches